src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -0,0 +1,444 @@
+
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.io.Files;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.lifecycle.LifecycleNewTracker;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.StorageAttachedIndex;
+import org.apache.cassandra.index.sai.disk.ColumnIndexWriter;
+import org.apache.cassandra.index.sai.disk.PerSSTableComponentsWriter;
+import org.apache.cassandra.index.sai.disk.io.IndexOutputWriter;
+import org.apache.cassandra.index.sai.memory.RowMapping;
+import org.apache.cassandra.index.sai.utils.IndexFileUtils;
+import org.apache.cassandra.io.sstable.Component;
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.util.FileHandle;
+import org.apache.cassandra.schema.CompressionParams;
+import org.apache.cassandra.utils.FBUtilities;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.util.IOUtils;
+
+public class IndexDescriptor
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    public static final String SAI_DESCRIPTOR = "SAI";
+
+    private static final String SEPARATOR = "-";
+    private static final String SAI_SEPARATOR = "+";
+    private static final String EXTENSION = ".db";
+    private static final String TMP_EXTENSION = ".tmp";
+
+    // Version AA - The final 6.8 file format
+    private static final String VERSION_AA_PER_SSTABLE_FORMAT = "SAI_%s.db";
+    private static final String VERSION_AA_PER_INDEX_FORMAT = "SAI_%s_%s.db";
+
+    public final Version version;
+    public final Descriptor descriptor;
+    public final Set<IndexComponent> perSSTableComponents = Sets.newHashSet();
+    public final Map<String, Set<IndexComponent>> perIndexComponents = Maps.newHashMap();
+    public final Map<IndexComponent, File> onDiskFileMap = Maps.newHashMap();
+    public final Map<IndexComponent, File> onDiskTemporaryFileMap = Maps.newHashMap();
+
+    private IndexDescriptor(Version version, Descriptor descriptor)
+    {
+        this.version = version;
+        this.descriptor = descriptor;
+    }
+
+    public static IndexDescriptor create(Descriptor descriptor)
+    {
+        Preconditions.checkArgument(descriptor != null, "Descriptor can't be null");
+
+        for (Version version : Version.ALL_VERSIONS)
+        {
+            IndexDescriptor indexDescriptor = new IndexDescriptor(version, descriptor);
+
+            if (indexDescriptor.fileFor(IndexComponent.GROUP_COMPLETION_MARKER).exists())
+            {
+                indexDescriptor.registerSSTable();
+                return indexDescriptor;
+            }
+        }
+        return new IndexDescriptor(Version.LATEST, descriptor);
+    }
+
+    public IndexDescriptor registerSSTable()
```

Comment:
This name confused me because I thought it had something to do with sstable itself, not per-SSTable-components. 
What about naming it `registerPerSSTableComponents` and similarly `registerPerIndexComponents`?